File: app/src/main/java/com/nollpointer/dates/practise/PractiseDetailsPickerAdapter.java
package com.nollpointer.dates.practise;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckedTextView;

import androidx.recyclerview.widget.RecyclerView;

import com.nollpointer.dates.R;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class PractiseDetailsPickerAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private CharSequence[] titles;
    private int mode;
    private int type = 0;
    private ArrayList<Integer> centuries = new ArrayList<>();
    private CheckedTextView previousPickedType;

    private CheckedTextView titleCheckedTextView;

    public static final int CENTURY = 1;
    public static final int TYPE = 0;

    private boolean isLocked = false;

    public PractiseDetailsPickerAdapter(CharSequence[] positions, int mode) {
        this.titles = positions;
        this.mode = mode;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view;
        if (mode == TYPE)
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_select_singlechoice, parent, false);
        else
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_select_multichoice, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, final int position) {

        CheckedTextView view = holder.itemView.findViewById(android.R.id.text1);
        view.setText(titles[position]);
        if (mode == TYPE) {
            if (position == type) {
                view.setChecked(true);
                previousPickedType = view;
            } else
                view.setChecked(false);
        } else {
            if (centuries.contains(position))
                view.setChecked(true);
            else
                view.setChecked(false);
        }

        view.setEnabled(!isLocked);
    }

    @Override
    public int getItemCount() {
        return titles.length;
    }

    public void setType(int type) {
        this.type = type;
    }

    public void setTitleCheckedTextView(CheckedTextView titleCheckedTextView) {
        this.titleCheckedTextView = titleCheckedTextView;
    }

    public void setCenturies(ArrayList<Integer> centuries) {
        this.centuries = centuries;
    }

    public int getType() {
        return type;
    }

    public List<Integer> getCenturies() {
        Collections.sort(centuries);
        Log.e("ADAPTER", "getCenturies: " + centuries.toString());
        return centuries;
    }

    public void setLocked() {
        isLocked = true;
    }

    public void makeRandomValues() {
        Random random = new Random(System.currentTimeMillis());
        type = random.nextInt(3);
        int size = random.nextInt(titles.length);
        centuries.clear();
        if (size == titles.length - 1) {
            for (int i = 0; i < titles.length; i++) {
                centuries.add(i);
            }

        }else {
            for (int i = 0; i < size; i++)
                centuries.add(random.nextInt(titles.length - 1));
        }

        if (mode == CENTURY)
            titleCheckedTextView.setChecked(size == titles.length - 1);

        notifyDataSetChanged();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        ViewHolder(final View view) {
            super(view);
            View textView = view.findViewById(android.R.id.text1);
            if (mode == TYPE)
                textView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {

                        CheckedTextView view = (CheckedTextView) v;
                        if (!view.isChecked()) {
                            view.setChecked(true);
                            previousPickedType.setChecked(false);
                            previousPickedType = view;
                            type = getAdapterPosition();
                        }
                    }
                });
            else
                textView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        CheckedTextView view = (CheckedTextView) v;
                        if (view.isChecked()) {
                            view.setChecked(false);
                            centuries.remove(Integer.valueOf(getAdapterPosition()));
                        } else {
                            view.setChecked(true);
                            centuries.add(getAdapterPosition());
                        }

                        titleCheckedTextView.setChecked(centuries.size() == 10);
                    }
                });
        }
    }
}

File: app/src/main/java/com/nollpointer/dates/practise/PractiseDetailsPickerAdapter.kt
package com.nollpointer.dates.practise

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckedTextView
import androidx.recyclerview.widget.RecyclerView
import com.nollpointer.dates.R
import java.util.*

class PractiseDetailsPickerAdapter(private val titles: Array<CharSequence>, private val mode: Int) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private lateinit var previousPickedType: CheckedTextView
    private lateinit var titleCheckedTextView: CheckedTextView

    var type = 0
    private var centuries = ArrayList<Int>()
    private var isLocked = false

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = if (mode == TYPE)
            LayoutInflater.from(parent.context).inflate(R.layout.card_select_singlechoice, parent, false)
        else
            LayoutInflater.from(parent.context).inflate(R.layout.card_select_multichoice, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val view = holder.itemView.findViewById<CheckedTextView>(android.R.id.text1)
        view.text = titles[position]
        if (mode == TYPE) {
            if (position == type) {
                view.isChecked = true
                previousPickedType = view
            } else view.isChecked = false
        } else {
            view.isChecked = centuries.contains(position)
        }
        view.isEnabled = !isLocked
    }

    override fun getItemCount(): Int {
        return titles.size
    }

    fun setTitleCheckedTextView(titleCheckedTextView: CheckedTextView) {
        this.titleCheckedTextView = titleCheckedTextView
    }

    fun setCenturies(centuries: ArrayList<Int>) {
        this.centuries = centuries
    }

    fun getCenturies(): List<Int> {
        centuries.sort()
        Log.e("ADAPTER", "getCenturies: $centuries")
        return centuries
    }

    fun setLocked() {
        isLocked = true
    }

    fun makeRandomValues() {
        val random = Random(System.currentTimeMillis())
        type = random.nextInt(3)
        val size = random.nextInt(titles.size)
        centuries.clear()
        if (size == titles.size - 1) {
            for (i in titles.indices) {
                centuries.add(i)
            }
        } else {
            for (i in 0 until size) centuries.add(random.nextInt(titles.size))
        }
        if (mode == CENTURY) titleCheckedTextView.isChecked = size == titles.size - 1
        notifyDataSetChanged()
    }

    inner class ViewHolder internal constructor(view: View) : RecyclerView.ViewHolder(view) {
        init {
            val textView = view.findViewById<View>(android.R.id.text1)
            textView.setOnClickListener {
                val checkedTextView = it as CheckedTextView
                if (mode == TYPE)
                    if (!checkedTextView.isChecked) {
                        checkedTextView.isChecked = true
                        previousPickedType.isChecked = false
                        previousPickedType = checkedTextView
                        type = adapterPosition
                    } else {
                        if (checkedTextView.isChecked) {
                            checkedTextView.isChecked = false
                            centuries.remove(Integer.valueOf(adapterPosition))
                        } else {
                            checkedTextView.isChecked = true
                            centuries.add(adapterPosition)
                        }
                        titleCheckedTextView.isChecked = centuries.size == 10
                    }

            }
        }
    }

    companion object {
        const val CENTURY = 1
        const val TYPE = 0
    }

}